File: src/main/java/io/vproxy/vfx/manager/image/ImageManager.java
package io.vproxy.vfx.manager.image;

import io.vproxy.base.util.LogType;
import io.vproxy.base.util.Logger;
import io.vproxy.vfx.util.FXUtils;
import javafx.scene.image.Image;

import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

public class ImageManager {
    private static final ImageManager instance = new ImageManager();

    public static ImageManager get() {
        return instance;
    }

    private ImageManager() {
        loadBlackAndChangeColor("io/vproxy/vfx/res/image/close.png", Map.of("red", 0xffed6a5e, "white", 0xffffffff));
        loadBlackAndChangeColor("io/vproxy/vfx/res/image/maximize.png", Map.of("white", 0xffffffff, "green", 0xff61c454));
        loadBlackAndChangeColor("io/vproxy/vfx/res/image/reset-window-size.png", Map.of("white", 0xffffffff, "green", 0xff61c454));
        loadBlackAndChangeColor("io/vproxy/vfx/res/image/iconify.png", Map.of("white", 0xffffffff, "yellow", 0xfff4bd4f));
        loadBlackAndChangeColor("io/vproxy/vfx/res/image/arrow.png", Map.of("white", 0xffffffff));
    }

    private final Map<String, Image> map = new ConcurrentHashMap<>();
    private final WeakHashMap<String, Image> weakMap = new WeakHashMap<>();

    public Image load(String path) {
        try {
            return load(path, false);
        } catch (Exception ignore) {
            return null;
        }
    }

    @SuppressWarnings("RedundantThrows")
    public Image load(String path, boolean throwException) throws Exception {
        if (!path.startsWith("/")) {
            path = "/" + path;
        }
        var image = map.get(path);
        if (image == null) {
            image = weakMap.get(path);
        }
        if (image != null) {
            assert Logger.lowLevelDebug("using cached image: " + path);
            return image;
        }
        try {
            image = new Image(path, false);
        } catch (Exception e) {
            Logger.error(LogType.FILE_ERROR, "failed loading image " + path, e);
            if (throwException) {
                throw e;
            }
            return null;
        }
        map.put(path, image);
        assert Logger.lowLevelDebug("new image loaded: " + path);
        return image;
    }

    public void loadBlackAndChangeColor(String path, Map<String, Integer> argbs) {
        if (!path.startsWith("/")) {
            path = "/" + path;
        }
        var img = load(path);
        if (img == null) {
            return;
        }
        for (var entry : argbs.entrySet()) {
            var name = entry.getKey();
            int setArgb = entry.getValue();
            var wImg = FXUtils.changeColorOfBlackImage(img, setArgb);
            var newPath = path + ":" + name;
            assert Logger.lowLevelDebug("new image loaded: " + newPath);
            map.put(newPath, wImg);
        }
    }

    public Image loadSubImageOrMake(String baseName, String subName, Function<Image, Image> makeFunc) {
        if (!baseName.startsWith("/")) {
            baseName = "/" + baseName;
        }
        var key = baseName + ":" + subName;
        var img = map.get(key);
        if (img == null) {
            img = weakMap.get(key);
        }
        if (img != null) {
            assert Logger.lowLevelDebug("using cached image: " + key);
            return img;
        }
        img = map.get(baseName);
        if (img == null) {
            img = weakMap.get(baseName);
        }
        if (img == null) {
            Logger.warn(LogType.ALERT, "unable to find base image " + baseName + ", cannot make sub image for it");
            return null;
        }
        img = makeFunc.apply(img);
        if (img == null) {
            Logger.warn(LogType.ALERT, "failed making image for " + key + ", the make function returns null");
            return null;
        }
        map.put(key, img);
        assert Logger.lowLevelDebug("new image loaded: " + key);
        return img;
    }

    public Image weakRef(String path) {
        var audio = map.remove(path);
        if (audio == null) {
            return null;
        }
        weakMap.put(path, audio);
        return audio;
    }

    public void remove(String path) {
        if (!path.startsWith("/")) {
            path = "/" + path;
        }
        if (!map.containsKey(path) && !weakMap.containsKey(path)) {
            return;
        }
        map.remove(path);
        weakMap.remove(path);
        assert Logger.lowLevelDebug("image removed: " + path);
    }
}

File: src-ts/src/manager/image/ImageManager.ts
import { Logger } from '../../vproxy-base/Logger.js';
import { FXUtils } from '../../util/FXUtils.js';
import { FXImage } from '../../javafx/ImageView.js';

const RES_BASE = './res/image/';

function toAssetPath(path: string): string {
  if (!path.startsWith('/')) path = '/' + path;
  const idx = path.indexOf('/io/vproxy/vfx/res/image/');
  if (idx >= 0) {
    return RES_BASE + path.substring(idx + '/io/vproxy/vfx/res/image/'.length);
  }
  if (path.startsWith('/io/vproxy/vfx/res/')) {
    return './res/' + path.substring('/io/vproxy/vfx/res/'.length);
  }
  return '.' + path;
}

export class ImageManager {
  static readonly instance = new ImageManager();

  static get(): ImageManager {
    return ImageManager.instance;
  }

  private readonly map = new Map<string, FXImage>();
  private readonly weakMap = new Map<string, WeakRef<FXImage>>();
  private readonly pending = new Map<string, Promise<void>>();
  private preloaded = false;

  private constructor() {
    this.loadBlackAndChangeColor('io/vproxy/vfx/res/image/close.png', {
      red: 0xffed6a5e, white: 0xffffffff,
    });
    this.loadBlackAndChangeColor('io/vproxy/vfx/res/image/maximize.png', {
      white: 0xffffffff, green: 0xff61c454,
    });
    this.loadBlackAndChangeColor('io/vproxy/vfx/res/image/reset-window-size.png', {
      white: 0xffffffff, green: 0xff61c454,
    });
    this.loadBlackAndChangeColor('io/vproxy/vfx/res/image/iconify.png', {
      white: 0xffffffff, yellow: 0xfff4bd4f,
    });
    this.loadBlackAndChangeColor('io/vproxy/vfx/res/image/arrow.png', {
      white: 0xffffffff,
    });
  }

  async preload(): Promise<void> {
    if (this.preloaded) return;
    await Promise.all([...this.pending.values()].map((p) => p.catch(() => undefined)));
    this.preloaded = true;
  }

  load(path: string): FXImage | null {
    if (!path.startsWith('/')) path = '/' + path;
    const cached = this.map.get(path);
    if (cached) return cached;
    const weak = this.weakMap.get(path)?.deref();
    if (weak) return weak;

    // Recolored paths ("foo.png:color") are produced asynchronously by
    // loadBlackAndChangeColor. If a recolored entry is requested before
    // its promise resolves, fall back to the base image.
    const colonIdx = path.indexOf(':');
    if (colonIdx >= 0) {
      const base = path.substring(0, colonIdx);
      const baseImg = this.map.get(base);
      if (baseImg) {
        const placeholder = new FXImage(baseImg.url);
        this.map.set(path, placeholder);
        const pending = this.pending.get(path);
        if (pending) {
          void pending.then(() => {
            const recolored = this.map.get(path);
            if (recolored && recolored !== placeholder) {
              placeholder.setUrl(recolored.url);
            }
          });
        }
        return placeholder;
      }
    }

    const assetPath = toAssetPath(path);
    const img = new FXImage(assetPath);
    this.map.set(path, img);
    return img;
  }

  loadBlackAndChangeColor(path: string, argbs: Record<string, number>): void {
    if (!path.startsWith('/')) path = '/' + path;
    const baseImg = this.load(path);
    if (baseImg === null) return;

    for (const [name, setArgb] of Object.entries(argbs)) {
      const newPath = path + ':' + name;
      const target = new FXImage('');
      this.map.set(newPath, newPath in this.map ? this.map.get(newPath)! : target);

      const promise = new Promise<void>((resolve) => {
        const baseHtml = baseImg.getHtmlImage();
        const start = () => {
          FXUtils.changeColorOfBlackImage(baseHtml, setArgb)
            .then((dataUrl) => {
              target.setUrl(dataUrl);
              resolve();
            })
            .catch((e) => {
              Logger.error('FILE_ERROR', `failed recoloring ${newPath}`, e);
              resolve();
            });
        };
        if (baseHtml.complete && baseHtml.naturalWidth > 0) {
          start();
        } else {
          baseHtml.addEventListener('load', start, { once: true });
          baseHtml.addEventListener('error', () => resolve(), { once: true });
        }
      });
      this.pending.set(newPath, promise);
    }
  }

  loadSubImageOrMake(
    baseName: string,
    subName: string,
    makeFunc: (img: FXImage) => FXImage,
  ): FXImage | null {
    if (!baseName.startsWith('/')) baseName = '/' + baseName;
    const key = baseName + ':' + subName;
    const cached = this.map.get(key);
    if (cached) return cached;
    const weak = this.weakMap.get(key)?.deref();
    if (weak) return weak;
    const base = this.map.get(baseName) ?? this.weakMap.get(baseName)?.deref();
    if (!base) {
      Logger.warn('ALERT', `unable to find base image ${baseName}, cannot make sub image for it`);
      return null;
    }
    const made = makeFunc(base);
    if (!made) {
      Logger.warn('ALERT', `failed making image for ${key}, the make function returns null`);
      return null;
    }
    this.map.set(key, made);
    return made;
  }

  weakRef(path: string): FXImage | null {
    const img = this.map.get(path);
    if (!img) return null;
    this.map.delete(path);
    this.weakMap.set(path, new WeakRef(img));
    return img;
  }

  remove(path: string): void {
    if (!path.startsWith('/')) path = '/' + path;
    if (!this.map.has(path) && !this.weakMap.has(path)) return;
    this.map.delete(path);
    this.weakMap.delete(path);
  }
}
